data/virt_autotest/guest_console_monitor.sh

```diff
@@ -0,0 +1,284 @@
+vm_guestnames_array=$(echo -e ${get_vm_guestnames})
+vmguest=""
+vmguest_failed="0"
+
+#Parse input arguments. -s or -e must have values
+while getopts "seh" OPTION; do
+  case "${OPTION}" in
+    s)
+      START_MONITOR="1"
+      ;;
+    e)
+      END_MONITOR="1"
+      ;;
+    h)
+      help_usage
+      exit 1
+      ;;
+    *)
+      help_usage
+      exit 1
+      ;;
+  esac
+done
+
+#FUNCTION PART
+function get_console()
+{
+
+local vmguest=$1
+
+expect -c "
+## -1 means never timeout
+set timeout -1
+
+##hide echo
+log_user 0
+spawn -noecho virsh console ${vmguest}
+
+#wait connection
+sleep 3
+send \"\r\n\r\n\r\n\"
+
+#condition expect
+expect {
+        \"*login:\" {
+                send \"root\r\"
+                exp_continue
+        }
+        \"*assword\" {
+                send \"novell\r\"
+                exp_continue
+        }
+        \"*:~ #\" {
+                send -- \"ip route get 1\r\"
+                exp_continue
+        }
+        \"*:~ #\" {
+                send -- \"exit\r\"
+                exp_continue                
+        }         
+}
+
+#submatch for output
+expect -re {dev.*\s([0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3})[^0-9]}
+
+if {![info exists expect_out(1,string)]} {
+        puts \"Match did not happen :(\"
+        exit 1
+}
+set output \$expect_out(buffer)
+expect eof
+
+puts \"\$output\"
+"
+
+}
+
+
+function open_console()
+{
+        #open guest console
+        for vmguest in ${vm_guestnames_state_running_array[@]};do
+                get_console ${vmguest} >> ${LOG_DIR}/${vmguest}_console_${timestamp} 2>&1 &
+        done 
+}
+
+function close_console()
+{
+        #close guest console
+        ##Kill parent PID from "basename $0" 
+        pid=`ps aux | grep "$(basename $0)" | grep -v "grep"| awk '{print $2}'`
+        kill -9 $pid >> /dev/null 2>&1
+        
+        ##Kill all associated "virsh cnsole" processes
+        for i in `seq 10`;do
+            ps aux | grep "virsh console" >> /dev/null 2>&1
+            if [ $? -eq 0 ];then
+                pid=`ps aux | grep "virsh console" | grep -v "grep"| awk '{print $2}'`
+                kill -9 $pid >> /dev/null 2>&1
+            else
+                break
+            fi
+        done
+}
+
+
+
+function end_monitor()
+{
+        #end of monitor for guest console
+		if [ -f ${MONITOR_LOCKED_FILE} ];then
+            echo "End of Monitor for guest console"		    
+            rm -rf ${MONITOR_LOCKED_FILE}
+            close_console
+        else
+            echo "There was not ${MONITOR_LOCKED_FILE}"
+            exit 1
+        fi
+}
+
+
+function post_fail_hook()
+{
+    end_monitor
+    exit 1
+}
+
+function start_monitor()
+{
+        #start monitor for guest console
+        get_vm_guestnames_state_running=`virsh list --state-running | grep -E "${vm_guestnames_types}" | awk '{print $2}'`
+        vm_guestnames_state_running_array=$(echo -e ${get_vm_guestnames_state_running})  
```

Comment:
vm state running array is only calculated once. Pay attention.

Reply:
sure,  here `vm state running array` just follow **`check_guest`** step, just only calculated once.

Reply:
sure, got it.